src/components/App/App.js

```diff
@@ -14,37 +14,140 @@ import RegSuccessPopup from '../RegSuccessPopup/RegSuccessPopup';
+  const [savedArticles, setSavedArticles] = useState([]);
+  const [savedCardsArray, setSavedCardsArray] = useState([]);
+
+  // Search articles from api by keyword
+  function handleSearchSubmit(keyword) {
+    setIsNewsCardListOpen(false);
+    setIsLoading(true);
+    newsApi
+      .getArticles(keyword)
       .then((res) => {
-        console.log('res', res);
-        closeAllPopups(true);
-        setRegSuccessPopupOpen(true);
-        setIsSignupPopupOpen(false);
+        setIsNewsCardListOpen(true);
+        setCards(res);
+        if (res.length === 0) {
+          setHasResults(false);
+        } else {
+          setHasResults(true);
+        }
+      })
+      .catch((err) => {
+        console.log(`Something went wrong: ${err}`);
+      })
+      .finally(() => {
+        setIsLoading(false);
+      })
+  }
+
+  //Save articles and adds it to an array of articles
+  function handleArticleSaving(data) {
+console.log(data,"app data check");
+    if (!savedArticles
+      .find((obj) => obj.title === data.title)) {
+      mainApi
+        .saveArticle(data, searchKeyword, jwt)
+        .then(() => {
+          mainApi.getSavedArticles(jwt)
+          .then((updatedArticles) => {
+            setSavedArticles(updatedArticles)
+          })
+        })
+        // .then((res) => {
+        //   if (res) {
+        //     setSavedArticles((prevSavedArticles) => [...prevSavedArticles, res.data]);
+        //   }
+        // })
+        .catch((err) => {
+          console.log(`Something went wrong: ${err}`);
+        })
+    } else {
+      console.log('you have allready saved this article');
+    }
+  }
 
+  // DELETE-ing article and removes it from the array
+  function handleRemoveArticle(data) {
+    let articleId;
+
+    if (isInHomepage) {
+      if (savedArticles.find((obj) => obj.link === data.url)) {
+        const article = savedArticles.find((obj) => {
+          return obj.link === data.url;
+        });
+        articleId = article._id;
+      } else {
+        console.log("this card does not exist");
+      }
+    } else {
+      articleId = data._id;
+    }
+    mainApi
+      .unsaveArticle(articleId, jwt)
+      .then(() => {
+        mainApi.getSavedArticles(jwt)
+        .then((updatedArticles) => {
+          setSavedArticles(updatedArticles)
+        })
       })
       .catch((err) => {
         console.log(`Something went wrong: ${err}`);
+      });
+      
+  }
 
+  function handleSignup({ email, password, name }) {
+    auth
+      .register(email, password, name)
+      .then((res) => {
+        if (res) {
+          setIsRegistered(true);
+          setRegSuccessPopupOpen(true);
+          setIsSignupPopupOpen(false);
+        } else {
+          setIsRegistered(false);
+        }
+      })
+      .catch((err) => {
+        console.log(`Something went wrong: ${err}`);
       });
   }
 
   function handleLogin(email, password) {
-    auth.authorize(email, password)
```

Comment:
You can create a special variable for watching **loading** states during server requests:

```js
 const [isLoading, setIsLoading] = React.useState(false);

```

And then you can change it to `true` before sending requests and return it back to `false`  in `finally` to stop loading:

And you should pass `isLoading` to every popup to change the **button text**:

```js

 buttonText={isLoading? 'Saving...' : 'Save'}

```